lib/segment/src/segment.rs

```diff
@@ -96,15 +96,51 @@ impl Segment {
     ) -> OperationResult<()> {
         debug_assert!(self.is_appendable());
         check_vectors_set(&vectors, &self.segment_config)?;
-        for (vector_name, vector) in vectors {
-            let vector_name: &str = &vector_name;
-            let vector_data = &self.vector_data[vector_name];
-            let mut vector_storage = vector_data.vector_storage.borrow_mut();
-            vector_storage.insert_vector(internal_id, &vector)?;
+        for (vector_name, vector_data) in self.vector_data.iter_mut() {
+            let vector = vectors.get(vector_name);
+            match vector {
+                Some(vector) => {
+                    let mut vector_storage = vector_data.vector_storage.borrow_mut();
+                    vector_storage.insert_vector(internal_id, vector)?;
+                }
+                None => {
+                    // No vector provided, so we remove it
+                    let mut vector_storage = vector_data.vector_storage.borrow_mut();
+                    vector_storage.delete_vec(internal_id)?;
+                }
+            }
         }
         Ok(())
     }
 
+    /// Insert new vectors into the segment.
+    /// WARN: Available for appendable segments only
+    fn insert_new_vectors(
+        &mut self,
+        point_id: PointIdType,
+        vectors: NamedVectors,
+    ) -> OperationResult<PointOffsetType> {
+        debug_assert!(self.is_appendable());
+        let new_index = self.id_tracker.borrow().internal_size() as PointOffsetType;
+        check_vectors_set(&vectors, &self.segment_config)?;
+        for (vector_name, vector_data) in self.vector_data.iter_mut() {
+            let vector_opt = vectors.get(vector_name);
+            let mut vector_storage = vector_data.vector_storage.borrow_mut();
+            match vector_opt {
+                None => {
+                    let dim = vector_storage.vector_dim();
+                    vector_storage.insert_vector(new_index, &vec![1.0; dim])?;
```

Comment:
Maybe we can do `insert_deleted_vector`, which would be less hacky. Or is complicating the storage trait not worth it in your opinion?